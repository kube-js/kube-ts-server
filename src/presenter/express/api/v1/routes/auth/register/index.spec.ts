import dotenv from 'dotenv';
dotenv.config();
import faker from 'faker';
import { CONFLICT, CREATED, UNPROCESSABLE_ENTITY } from 'http-status-codes';
import supertest from 'supertest';
import { TEXT_LENGTH } from '../../../../../../../constants';
import { API_V1, AUTH, REGISTER } from '../../../../../../../constants/routes';
import usersFactory from '../../../../../utils/fakeFactories/users/factory';
import initTests from '../../../../../utils/tests/initTests';

import {
  TEST_DIFFERENT_VALID_PASSWORD,
  TEST_VALID_EMAIL,
  TEST_VALID_PASSWORD,
} from '../../../../../utils/tests/testData';

const emailSubject = 'subject';
const link = 'link';

jest.mock('uuid', () => ({
  v4: jest.fn(() => '1'),
}));

jest.mock('../../../../../../../utils/helpers/url/getVerifyEmailUrl', () =>
  jest.fn(() => link)
);

jest.mock('../../../../../../../utils/helpers/date/getUtcDate', () =>
  jest.fn(() => '2000-04-05T23:26:42.000Z')
);

jest.mock('../../../../../../../utils/helpers/auth/generateToken', () =>
  jest.fn(() => 'token')
);

const validRequiredFields = {
  email: TEST_VALID_EMAIL,
  password: TEST_VALID_PASSWORD,
  password_confirmation: TEST_VALID_PASSWORD,
};

const REGISTER_URL = `${API_V1}${AUTH}${REGISTER}`;
interface AssertOptions {
  readonly request: supertest.SuperTest<any>;
  readonly fields?: { [key: string]: any };
  readonly statusCode?: number;
}

export const assertOnResponseAndStatus = async ({
  request,
  fields,
  statusCode = UNPROCESSABLE_ENTITY,
}: AssertOptions) => {
  const { status, body } = await request.post(REGISTER_URL).send(fields);

  expect(status).toBe(statusCode);
  expect(body).toMatchSnapshot();

  return { status, body };
};

export const assertWithRequiredFieldAlreadyIncluded = ({
  request,
  fields,
  statusCode,
}: AssertOptions) =>
  assertOnResponseAndStatus({
    fields: {
      ...validRequiredFields,
      ...fields,
    },
    request,
    statusCode,
  });

describe('@presenter/auth/register', () => {
  const { request, service, mailServer } = initTests({ useMailServer: true });

  afterAll(() => {
    jest.clearAllMocks();
  });

  it('fails to register user without required fields', async () => {
    await assertOnResponseAndStatus({
      request,
    });
  });

  it('fails to register user when email is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        password: TEST_VALID_PASSWORD,
        password_confirmation: TEST_VALID_PASSWORD,
      },
      request,
    });
  });

  it('fails to register user when password is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password_confirmation: TEST_VALID_PASSWORD,
      },
      request,
    });
  });

  it('fails to register user when password_confirmation is missing out of required properties', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password: TEST_VALID_PASSWORD,
      },
      request,
    });
  });

  it('fails to register user when password and password_confirmation are not matching', async () => {
    await assertOnResponseAndStatus({
      fields: {
        email: TEST_VALID_EMAIL,
        password: TEST_VALID_PASSWORD,
        password_confirmation: TEST_DIFFERENT_VALID_PASSWORD,
      },
      request,
    });
  });

  it('fails to register user when bio exceeds number of characters', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: {
        bio: faker.random.alphaNumeric(TEXT_LENGTH),
      },
      request,
    });
  });

  it('fails to register user when date_of_birth is not valid date', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { date_of_birth: 'invalid date' },
      request,
    });
  });

  it('fails to register user when first_name is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { first_name: 3222 },
      request,
    });
  });

  it('fails to register user when gender is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { gender: 'non-existing-gender' },
      request,
    });
  });

  it('fails to register user when last_name is invalid', async () => {
    await assertWithRequiredFieldAlreadyIncluded({
      fields: { last_name: 3232 },
      request,
    });
  });

  it('fails to register user when user with the email already exist', async () => {
    await usersFactory({
      overrides: {
        email: validRequiredFields.email,
      },
      service: service.users,
    });

    await assertWithRequiredFieldAlreadyIncluded({
      request,
      statusCode: CONFLICT,
    });
  });

  it('registers user returning back it visible fields and jwt token', async () => {
    jest.mock('../../../../../../../translator/factory.ts', () => () => ({
      verifyYourEmailHtml: jest.fn(() => link),
      verifyYourEmailSubject: () => emailSubject,
      verifyYourEmailText: jest.fn(() => link),
    }));

    const fields = {
      bio: 'short bio',
      date_of_birth: '1970-01-01',
      first_name: 'John',
      gender: 'male',
      last_name: 'Doe',
    };

    await assertWithRequiredFieldAlreadyIncluded({
      fields,
      request,
      statusCode: CREATED,
    });

    const messages = mailServer.messages;

    expect(messages.length).toBe(1);

    const { textContent, subject, to, from, htmlContent } = messages[0];

    expect(from).toMatchSnapshot('from');
    expect(htmlContent).toMatchSnapshot('htmlContent');
    expect(textContent).toMatchSnapshot('textContent');
    expect(subject).toMatchSnapshot('subject');
    expect(to).toMatchSnapshot('to');
  });
  // tslint:disable-next-line:max-file-line-count
});
