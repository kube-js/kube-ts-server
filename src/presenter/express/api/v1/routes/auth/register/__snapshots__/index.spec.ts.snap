// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@presenter/auth/register fails to register user when bio exceeds number of characters 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "bio",
      "message": "Expected string with a length between 0 and 65535 characters",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when date_of_birth is not valid date 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "date_of_birth",
      "message": "Expected date in one of the formats: DD-MM-YYYY, YYYY-MM-DD",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when email is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Invalid email",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when first_name is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "first_name",
      "message": "Expected string with a length between 0 and 255 characters",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when gender is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "gender",
      "message": "Expected one of the following: female, male",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when last_name is invalid 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "last_name",
      "message": "Expected string with a length between 0 and 255 characters",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when password and password_confirmation are not matching 1`] = `
Object {
  "errors": Array [
    Object {
      "fields": Array [
        "password",
        "password_confirmation",
      ],
      "message": "Expected password to match password_confirmation",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when password is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "password",
      "message": "Expected string minimum 8 characters long, containing at least: 1 upper and 1 lower case, 1 digit and 1 special characters i.e. one of the following: #?!@$%^&*-",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when password_confirmation is missing out of required properties 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "password_confirmation",
      "message": "Expected string minimum 8 characters long, containing at least: 1 upper and 1 lower case, 1 digit and 1 special characters i.e. one of the following: #?!@$%^&*-",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register fails to register user when user with the email already exist 1`] = `
Object {
  "message": "Model of type User already exists",
}
`;

exports[`@presenter/auth/register fails to register user without required fields 1`] = `
Object {
  "errors": Array [
    Object {
      "field": "email",
      "message": "Invalid email",
    },
    Object {
      "field": "password",
      "message": "Expected string minimum 8 characters long, containing at least: 1 upper and 1 lower case, 1 digit and 1 special characters i.e. one of the following: #?!@$%^&*-",
    },
    Object {
      "field": "password_confirmation",
      "message": "Expected string minimum 8 characters long, containing at least: 1 upper and 1 lower case, 1 digit and 1 special characters i.e. one of the following: #?!@$%^&*-",
    },
  ],
  "message": "Validation failed",
}
`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token 1`] = `
Object {
  "token": "token",
  "user": Object {
    "bio": "short bio",
    "date_of_birth": "1969-12-31T23:00:00.000Z",
    "email": "valid.email@test.com",
    "first_name": "John",
    "gender": "male",
    "id": "1",
    "last_name": "Doe",
  },
}
`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: from 1`] = `"noreply@cognitio.pl"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: htmlContent 1`] = `"<h1>Hello,</h1><p>To activate your account click <a href=\\"link\\">here</a>, alternatively if the address is not active copy/paste into your browser.</p><p>kube-ts-server</p>"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: subject 1`] = `"[kube-ts-server] Verify your email"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: textContent 1`] = `"Hello, to activate your account click: link, alternatively if the address is not active copy/paste into your browser. kube-ts-server"`;

exports[`@presenter/auth/register registers user returning back it visible fields and jwt token: to 1`] = `"valid.email@test.com"`;
